pkg/clustermesh/kvstoremesh/kvstoremesh.go

```diff
@@ -165,6 +174,9 @@ func (km *KVStoreMesh) newRemoteCluster(name string, status common.StatusFunc) c
 
 	run(rc.nodes.syncer.Run)
 	run(rc.services.syncer.Run)
+	if km.mcsAPIConfig.ClusterMeshEnableMCSAPI {
```

Comment:
Shouldn't this logic and code-base be entirely contained with in the `pkg/clustermesh/mcsapi` cell? This seems (at first glance) like a bit of a leak.

Reply:
Hmm not sure we can do much better than this tbh :/. This is all code coming from the rest of the `kvstoremesh` and is very similar to what we are doing for nodes, services and so on. The only thing that is specific to mcsapi is that we have a config to disable it but apart from that it's pretty much the same. However if you or anyone else have some suggestions/tips to improve this I am happy to check if it can be improved and do it if feasible!

Reply:
Hmm, I'd actually suggest to simplify this logic and always blindly attempt to synchronize this prefix in kvstoremesh. If not enabled, then nothing will need to be synchronized, so the end result would be the same in any case.

Reply:
I actually tried that initially but afaik the kvstoremesh was never ready if it was trying to sync service export while we were not exporting them, I didn't digged the exact reason though but I can check more

Reply:
I guess this is because the clustermesh-apiserver (or the operator) didn't write the "synced" flag for that resource type. One possibility could be just always writing it (i.e., starting the syncstore regardless of whether it is enabled, and calling `Synced`.

Reply:
I did that, it works locally and the code looks better thanks! There is one issue though, the ci-clustermesh upgrade tests are not happy :(.

I am suspecting that it may be related to the fact that this is an upgrade test, so with one cluster not running this version of the code. What doesn't make sense though is that the failing task "waiting for cluster mesh status to be ready" says that the pod clustermesh-apiserver is started but not ready but in the pods dumps it is ready 😭. Eitherway I will try to reproduce locally this kind of upgrade scenario. If it is related to upgrade I am suspecting that it may be kvstoremesh trying to pull service exports from acluster that does not send the "synced" flag (because on an older version).

Reply:
> If it is related to upgrade I am suspecting that it may be kvstoremesh trying to pull service exports from acluster that does not send the "synced" flag (because on an older version).

Right, this is most likely the case, and kvstoremesh (as well as the agents, when that part will be implemented), need to tolerate the possibility that the serviceexport information is not yet present to support connecting to older versions. If you need to share whether serviceexports are present/enabled or not, then the [Capabilities section of the CiliumClusterConfig](https://github.com/cilium/cilium/blob/43642e5cda6eeb8c7ab7af9d64579ce725501e85/pkg/clustermesh/types/types.go#L80-L97) should be the right place for it.

Reply:
Many thanks for the extra pointer, so I implemented this change and could get a setup like this to work locally but the CI is still not happy with it :/. Trying to dig why exactly

Reply:
Ok so the CI issue seems be specifically related to the kvstoremesh commit (https://github.com/cilium/cilium/pull/32972/commits/bc0c2f7784969a1039bc5a9a52c80b88b3d9ccff).

I think I need to start the reflector (newReflector + run method on the syncer) only if the remote cluster support service exports. But... I don't have the cluster capability available in `newRemoteCluster`. So three proposals:
1) Bring back the flag `ClusterMeshEnableMCSAPI` + remember to put a big warning in the doc to not activate mcs api support unless every clusters in the mesh has the cilium version necessary to run this (and perhaps keep the cluster capability but only as a mean to display an error log).

2) Have a way to add the service export reflector started in the `Run` method of the RemoteCluster

3) Refactor the RemoteCluster so that we can init all the reflectors at a time we have cluster capabilities available

WDYT @giorio94 :pray: ?

Reply:
From an initial look, I think that the missing part would be pretending that the synchronization completed in case service exports are disabled according to the capabilities (as it then eventually calls `o.synced.Done()`, which unblocks the readiness of kvstoremesh):

```diff
if srccfg.Capabilities.ServiceExportsSupport {
	serviceExportsSupported = true
	mgr.Register(adapter(mcsapitypes.ServiceExportStorePrefix), func(ctx context.Context) {
		rc.serviceExports.watcher.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
	})
+} else {
+	rc.serviceExports.syncer.OnSync(ctx)
+}
```

Another possibility may be manually calling `o.synced.Done()` only, to not add the synced key in the kvstore (depending on which approach we want to adopt there). However, if that's the case, care must be payed to only call it once, and not for every reconnection. 

Additionally, it seems appropriate to me to call `rc.serviceExports.watcher.Drain()` as well, to ensure that cached entries are deleted upon reconnection if the `ServiceExportsSupport` value changes.

More in general, the reason why the reflectors are immediately started is that their lifecycle is independent of the connection to the kvstore itself, which can be performed multiple times. Additionally, the capabilities are not guaranteed to remain consistent (e.g., in case of upgrade), and possible changes of the capabilities need to be handled as well. 

(By the way, thanks for splitting the commits. IMO the PR is way easier to read now :rocket: )